File: day_3/puzzle_1/solve.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const benchmark_1 = require("../../benchmark/benchmark");
(0, benchmark_1.benchmark)(() => {
    function letterToPriority(letter) {
        const charCode = letter.charCodeAt(0);
        return charCode >= 97 ? charCode - 96 : charCode - 38;
    }
    const file = fs_1.default.readFileSync("./day_3/input.txt", { encoding: "utf-8" });
    const ruckSacks = file.split("\n");
    const priority = ruckSacks.reduce((sum, ruckSack) => {
        const compartmentSize = ruckSack.length / 2;
        const compartmentOne = ruckSack.slice(0, compartmentSize);
        const compartmentTwo = ruckSack.slice(compartmentSize, ruckSack.length);
        for (const char of compartmentOne) {
            if (compartmentTwo.includes(char)) {
                sum += letterToPriority(char);
                break;
            }
        }
        return sum;
    }, 0);
    return priority;
}, "day_3_puzzle_1");

File: day_3/puzzle_2/solve.py
import os
import sys
from functools import reduce
from typing import List

sys.path.append(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
)

from benchmark.benchmark import benchmark


def main():
    def group_ruck_sacks(ruck_sacks: List[List[str]], size: int):
        return [ruck_sacks[i : i + size] for i in range(0, len(ruck_sacks), size)]

    def letter_to_priority(letter: str):
        char_code = ord(letter)
        return char_code - 96 if char_code >= 97 else char_code - 38

    def assess_group(sum: int, ruck_sacks: List[str]):
        for char in ruck_sacks[0]:
            if char in ruck_sacks[1] and char in ruck_sacks[2]:
                sum += letter_to_priority(char)
                break

        return sum

    with open("./day_3/input.txt", "r") as file:
        ruck_sack_groups = group_ruck_sacks(file.read().splitlines(), 3)
        priority = reduce(assess_group, ruck_sack_groups, 0)
        return priority


if __name__ == "__main__":
    benchmark(main, "day_3_puzzle_2")
